services/queue/src/api.js

```diff
@@ -309,6 +309,117 @@ builder.declare({
   return res.reply(result);
 });
 
+const cancelSingleTask = async (task, ctx) => {
+  // Get the last run, there should always be one
+  let run = _.last(task.runs);
+  if (!run) {
+    let err = new Error('There should exist a run after cancelTask!');
+    err.taskId = task.taskId;
+    err.status = task.status();
+    ctx.monitor.reportError(err);
+  }
+
+  // Construct status object
+  let status = task.status();
+
+  // If the last run was canceled, resolve dependencies and publish message
+  if (run.state === 'exception' && run.reasonResolved === 'canceled') {
+    // Update dependency tracker
+    await ctx.queueService.putResolvedMessage(
+      task.taskId,
+      task.taskGroupId,
+      task.schedulerId,
+      'exception',
+    );
+
+    // Publish message about the exception
+    const runId = task.runs.length - 1;
+    await ctx.publisher.taskException(_.defaults({
+      status,
+      runId,
+    }, _.pick(run, 'workerGroup', 'workerId')), task.routes);
+    ctx.monitor.log.taskException({ taskId: task.taskId, runId });
+  }
+
+  return status;
+};
+
+/** Cancel all tasks in a group */
+builder.declare({
+  method: 'post',
+  route: '/task-group/:taskGroupId/cancel',
+  name: 'cancelTaskGroup',
+  scopes: 'queue:cancel-task-group:<schedulerId>/<taskGroupId>',
```

Comment:
After some few other discussions we decided to require all of the scopes:

```
queue:seal-task-group:scheduler-id:<schedulerId>
queue:seal-task-group:task-group-id:<taskGropuId>
queue:seal-task-group:project-id:<projectId>
```

This way if someone wants to have flexibility, role with few scopes using `:*` can be used while some scope will be strict.

Reply:
@djmitche here is a brainstorming of how ideal scope should look like :)

Reply:
Per https://docs.taskcluster.net/docs/manual/tasks/groups, schedulerId _can_ be used to control who can manipulate tasks, but projectId is preferred (in fact, projectId was invented because schedulerId was doing too many things and impractical for this purpose).

That said, I think it makes sense to make this method parallel to [queue.cancelTask](https://docs.taskcluster.net/docs/reference/platform/queue/api#cancelTask):

any of
* queue:cancel-task:<schedulerId>/<taskGroupId>/<taskId>; or
* queue:cancel-task-in-project:<projectId>;
* all of
  * queue:cancel-task; and
  * assume:scheduler-id:<schedulerId>/<taskGroupId>

The third of those ("all of..") is deprecated, so no need to duplicate that. You've said you'll introduce project-id support later, which will correspond to the second. So, I think it makes sense to mirror the first option here, which is option 1 above.

This allows a few things:
 * `queue:$verb-task-group:*` -> cancel/seal anything you want
 * `queue:$verb-task-group:tc-github/*` -> cancel/seal tasks created by tc-github
 * `queue:$verb-task-group:$schedulerId/$taskGroupId` - permission to cancel this specific task group. This might be granted in temporary credentials, for example.

Reply:
Thanks Dustin!

I wonder if we should already add project specific scopes now or do in the follow up PR